[feature] partition prune support of list partition

hiliuxg opened the pull request:
issue #4035

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Problem Summary(Required) ：
olap table partition prune support of list partition

hiliuxg committed:
partition prune support of list partition

hiliuxg committed:
partition prune support of list partition

hiliuxg committed:
partition prune support of list partition

hiliuxg committed:
partition prune support of list partition

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 90 / 99 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java|7|10|70.00%|[88, 89, 91]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|7|10|70.00%|[252, 253, 255]|
|:large_blue_circle:|com/starrocks/planner/ListPartitionForOlapPruner.java|76|79|96.20%|[36, 47, 70]|

wyb commented:
I think you can use the current `ListPartitionPruner` directly.

hiliuxg committed:
reuse listPartitionPrune to prune the list partition data

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:
```diff
@@ -72,6 +85,69 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
                 input.getInputs()));
     }
 
+    private void putValueMapItem(TreeMap<LiteralExpr, Set<Long>> columnToPartitionValuesMap,
```
```suggestion
    private void putValueMapItem(TreeMap<LiteralExpr, Set<Long>> partitionValueToIds,
```

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:
```diff
@@ -72,6 +85,69 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
+        // single item list partition has only one column mapper
+        Map<Long, List<LiteralExpr>> literalExprValuesMap = listPartitionInfo.getLiteralExprValues();
+        if (literalExprValuesMap != null && literalExprValuesMap.size() > 0) {
+            TreeMap<LiteralExpr, Set<Long>> columnToPartitionValuesItem = new TreeMap<>();
```
```suggestion
            TreeMap<LiteralExpr, Set<Long>> partitionValueToIds = new TreeMap<>();
```

hiliuxg committed:
1. delete the unuseful file olapScanNodeTest.java
2. rename the variable name from columnToPartitionValuesItem to partitionValueToIds

Seaven commented on fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRuleTest.java:
```diff
@@ -32,13 +36,15 @@
 import static org.junit.Assert.assertNull;
 
-public class PartitionPruneRuleTest {
+public class PartitionPruneRuleTest{
```
```suggestion
public class PartitionPruneRuleTest {
```

Seaven commented:
reslove conflicts plz~

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:
```diff
@@ -72,6 +85,69 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
+        return null;
+    }
+
     private List<Long> partitionPrune(OlapTable olapTable, RangePartitionInfo partitionInfo,
```
```suggestion
    private List<Long> rangePartitionPrune(OlapTable olapTable, RangePartitionInfo partitionInfo,
```

hiliuxg committed:
rename some method
